crypto/x509/x509_lu.c

```diff
@@ -740,11 +740,7 @@ int X509_STORE_CTX_get1_issuer(X509 **issuer, X509_STORE_CTX *ctx, X509 *x)
     if (ctx->check_issued(ctx, x, obj->data.x509)) {
         if (x509_check_cert_time(ctx, obj->data.x509, -1)) {
             *issuer = obj->data.x509;
-            if (!X509_up_ref(*issuer)) {
-                *issuer = NULL;
-                ok = -1;
-            }
-            X509_OBJECT_free(obj);
+            OPENSSL_free(obj);
```

Comment:
The `X509_STORE_CTX_get_by_subject()` called before already does the up_ref on the cert using `X509_OBJECT_up_ref_count()`.
Then `X509_STORE_CTX_get1_issuer()` did an additional up_ref, but only because the following `X509_OBJECT_free()` did a down-ref.
This commit simply short-cuts the needless intermediate up&down-ref on the cert.

Reply:
OK, please make a comment to that effect here, and please consider updating the docs of `X509_STORE_set_get_issuer(3)` as suggested.

Reply:
This is truly highly confusing. I'd prefer keeping the code as is.

Reply:
Done as @vdukhovni suggested.

Also documented the default implementation: `X509_STORE_CTX_get1_issuer()`.

Reply:
@t8m, given the new comment I believe it should be clear now.

(The up&down ref also had the risk of needless failure.)

Reply:
Maybe more clear code would be something like:
```
/* *issuer is taking over the reference from the obj */
obj->type = X509_LU_NONE;
X509_OBJECT_free(obj);
```

Reply:
Good thought - done.